### PR TITLE
Fixes #5515

### LINKED ISSUE #5515: blameable behavior only for non-guests
Hi,
I've noticed that BlameableBehavior updates the updated_by field when a user is requesting a password reset.  As a result, updated_by becomes NULL as this is an un-recognised user (isGuest).
I'm thinking that BlameableBehavior should only execute if user is !isGuest() .
Not sure if there other instances where BlameableBehavior can be influenced by isGuest, I'm new to yii in general, so starting with yii2.

### PR DESCRIPTION
**Pull request by dmirogin**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #5515

### CODE REVIEW

**Comment by cebe** [comment]
ideally the behavior should be disabled when user is guest to not override a value set from a real user.

File: framework/behaviors/BlameableBehavior.php [hunk 1]
```diff
@@ -73,6 +73,15 @@ class BlameableBehavior extends AttributeBehavior
...
 
+    /**
+     * @var mixed Default value for cases when user is guest
+     */
+    public $defaultValue;
+
+    /**
+     * @var bool Whether is [[defaultValue]] enabled
+     */
+    public $useDefaultValue = false;
```
Comment (rob006) [inline framework/behaviors/BlameableBehavior.php]: This property is redundant. `$defaultValue` is `null` by default, same as `Yii::$app->get('user')->id` for guests. We don't need separate property to decide which `null` should we use.

File: framework/behaviors/BlameableBehavior.php [hunk 2]
```diff
@@ -97,7 +106,12 @@ public function init()
     protected function getValue($event)
     {
         if ($this->value === null && Yii::$app->has('user')) {
-            return Yii::$app->get('user')->id;
+            $userId = Yii::$app->get('user')->id;
+            if (is_null($userId) && $this->useDefaultValue) {
```
Comment (rob006) [inline framework/behaviors/BlameableBehavior.php]: `is_null($userId)` => `$userId !== null`

File: framework/behaviors/BlameableBehavior.php [hunk 3]
```diff
@@ -73,6 +73,15 @@ class BlameableBehavior extends AttributeBehavior
      */
     public $value;
 
+    /**
+     * @var mixed Default value for cases when user is guest
+     */
+    public $defaultValue;
```
Comment (rob006) [inline framework/behaviors/BlameableBehavior.php]: Missing `@since`.

**Review by rob006 — CHANGES_REQUESTED** [review]

File: framework/behaviors/BlameableBehavior.php (hunk 2 above)
Comment (dmirogin) [reply framework/behaviors/BlameableBehavior.php]: @rob006 http://php.net/manual/ru/function.is-null.php#87355

Comment (cebe) [reply framework/behaviors/BlameableBehavior.php]: that comment does not include `===`. behaviour equal and `===` is consistent with other yii code,

**Comment by dmirogin** [comment]
I understood completely this case, i'll finish in this week.

**Comment by samdark** [comment]
@dmirogin any update?

**Comment by dmirogin** [comment]
Sorry, was busy, going to finish in this weekend, maybe earlier.

**Comment by dmirogin** [comment]
i've also noticed a lot of similar constructions for determining closures, like
~~~php
if ($value instanceof \Closure || (is_array($value) && is_callable($value))) {
     return call_user_func($value);
}
~~~

Would be great to add this function inside some helper

**Comment by dmirogin** [comment]
@samdark Could you take a look?

**Review by samdark — APPROVED** [review]

**Review by rob006 — APPROVED** [review]

**Review by sergeymakinen — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**